Ignore raw `pre_save` signal

timobrembeck opened the pull request:
This adds support for the `raw` flag to linkcheck's `pre_save` signal.

If the model is saved exactly as presented (i.e. when loading a fixture), one should not query/modify other records in the database as the database might not be in a consistent state yet (see [documentation](https://docs.djangoproject.com/en/3.2/ref/signals/#pre-save)).

This is my first contribution to this repository, so I'm not sure how you would want this change to be represented in the tests. But if you give me a hint, I'm willing to add test cases for this patch.
Thanks in advance for your feedback and help!

Fixes  #106

timobrembeck committed:
Ignore raw pre_save signal (fixes  #106)

If the model is saved exactly as presented (i.e. when loading a fixture), one should not query/modify other records in the database as the database might not be in a consistent state yet.

claudep commented:
Hi Timo, thanks for the report and the patch!
Do you think you would be able to add a fixture to the tests, containing a model like yours that fails in get_absolute_url?

timobrembeck committed:
Add test case for fixtures

Load a fixture which has an id already - this way, the instance_pre_save() function is executed but fails because it relies on the fact that the object instance exists in the database already.

timobrembeck commented:
@claudep I now added a test case which fails for the current master branch and is fixed with my patch applied.

This was much simpler than I thought because there is another problem with fixtures independent from my special `get_absolute_url()`: The function `instance_pre_save()` tries to query the previous object from the database, but in the case of fixtures, this object does not yet exists, even though it has an id:
https://github.com/DjangoAdminHackers/django-linkcheck/blob/d086a1ef2d8db56c6fe3f4e49f80238b9bcce782/linkcheck/listeners.py#L119
Both problems are fixed if the pre-save method ignores `raw` signals.

claudep commented on linkcheck/listeners.py:
```diff
@@ -111,9 +111,9 @@ def delete_instance_links(sender, instance, **kwargs):
-def instance_pre_save(sender, instance, **kwargs):
-    if not instance.pk:
-        # Ignore unsaved instances
+def instance_pre_save(sender, instance, raw, **kwargs):
```
I have a slight preference to using `raw=False` in the signature (technically, this is a kwarg).

claudep commented on linkcheck/tests/test_linkcheck.py:
```diff
@@ -416,3 +416,10 @@ def test(self):
+    fixtures = ['linkcheck/tests/sampleapp/fixture.json']
+
+    def test(self):
+        pass
```
It wouldn't hurt to assert here the presence of the Book instance.

timobrembeck replied:
Alright! Like this?

claudep replied:
I guess the first assertion (count = 1) is sufficient.

timobrembeck replied:
Ok, I removed the other assertions.

claudep reviewed (changes requested):
Thanks a lot! Just two minor suggestions for the patch.

claudep commented:
Thanks a lot!
